icubam/messaging/scheduler.py

```diff
@@ -1,74 +1,103 @@
+  timeout: object = None
+  msg: message.Message = None
+  when: int = -1
+
+
 class MessageScheduler:
   """Schedules the sending of SMS to users."""
 
-  def __init__(self,
-               config,
-               db,
-               queue,
-               token_encoder,
-               base_url: str = 'http://localhost:8888/',
-               max_retries: int = 2,
-               reminder_delay: int = 60*30,
-               when=[(9, 30), (17, 0)]):
+  def __init__(self, config, db, queue):
     self.config = config
     self.db = db
-    self.token_encoder = token_encoder
     self.queue = queue
-    self.base_url = base_url
-    self.max_retries = max_retries
-    self.reminder_delay = reminder_delay
-    self.when = [time_utils.parse_hour(h) for h in when]
-    self.phone_to_icu = {}
-    self.messages = []
+    self.base_url = self.config.server.base_url
+    self.max_retries = self.config.scheduler.max_retries
+    self.reminder_delay = self.config.scheduler.reminder_delay
+    pings = self.config.scheduler.ping
+    self.when = [time_utils.parse_hour(h) for h in pings]
+
+    # Keys: by (user_id, icu_id), value is ScheduledMessage
     self.timeouts = {}
     self.updater = updater.Updater(self.config, None)
-    self.build_messages()
 
-  def build_messages(self):
-    """Build the messages to be sent to each user depending on its ICU."""
+  def computes_delay(self, delay=None) -> int:
+    """Computes the delay if None."""
+    if delay is not None:
+      return int(delay)
+
+    if not self.when:
+      logging.warning('No ping time. Check config.')
+      return -1
+
+    return int(time_utils.get_next_timestamp(self.when) - time.time())
+
+  def schedule_message(
+      self, msg: message.Message, delay: Optional[int]) -> bool:
+    """Schedules a message to be sent later."""
+    delay = self.computes_delay(delay)
+    if delay < 0:
+      logging.error("Negative delay for scheduling: skipping")
+      return False
+
+    when = delay + time.time()
+    key = msg.user_id, msg.icu_id
+    timeout = self.timeouts.get(key, None)
+    if timeout is not None and when > timeout.when:
+      logging.info(f'A message is schedule before {when}, Skipping scheduling.')
+      return False
+
+    io_loop = tornado.ioloop.IOLoop.current()
+    if timeout is not None:
+      self.unschedule(msg.user_id, msg.icu_id)
+
+    handle = io_loop.call_later(delay, self.may_send, msg)
+    self.timeouts[key] = ScheduledMessage(handle, msg, when)
+    logging.info('Scheduling {} in {}s.'.format(msg.icu_name, delay))
+    return True
+
+  def schedule(self, user, icu_id: int, delay: Optional[int] = None) -> bool:
+    user_icus = {i.icu_id: i for i in user.icus}
+    if not user.is_active or icu_id not in user_icus:
+      user_id = user.user_id
+      logging.info(f'Cannot send message to user {user_id} in icu {icu_id}')
+      return False
+
+    url = self.updater.get_user_url(user, icu_id)
+    msg = message.Message(icu_id, user, url)
```

Comment:
Since you are changing Message anyway, it may be cleaner if we pass the icu instead of icu_id here. That will also simplify the checks there.